File: library/src/main/java/com/thesurix/gesturerecycler/GestureManager.java
package com.thesurix.gesturerecycler;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

/**
 * Class that is responsible for gesture management for {@link RecyclerView} widget.
 * @author thesurix
 */
public class GestureManager {

    private GestureTouchHelperCallback mTouchHelperCallback;

    private GestureManager(final Builder builder) {
        final GestureAdapter adapter = (GestureAdapter) builder.recyclerView.getAdapter();
        mTouchHelperCallback = new GestureTouchHelperCallback(adapter);
        mTouchHelperCallback.setSwipeEnabled(builder.isSwipeEnabled);
        mTouchHelperCallback.setLongPressDragEnabled(builder.isDragEnabled);
        mTouchHelperCallback.setManualDragEnabled(builder.isManualDragEnabled);

        final ItemTouchHelper touchHelper = new ItemTouchHelper(mTouchHelperCallback);
        touchHelper.attachToRecyclerView(builder.recyclerView);
        adapter.setGestureListener(new GestureListener(touchHelper));

        if (builder.swipeFlags == Builder.INVALID_FLAG) {
            mTouchHelperCallback.setSwipeFlagsForLayout(builder.recyclerView.getLayoutManager());
        } else {
            mTouchHelperCallback.setSwipeFlags(builder.swipeFlags);
        }

        if (builder.dragFlags == Builder.INVALID_FLAG) {
            mTouchHelperCallback.setDragFlagsForLayout(builder.recyclerView.getLayoutManager());
        } else {
            mTouchHelperCallback.setDragFlags(builder.dragFlags);
        }
    }

    /**
     * Sets swipe gesture enabled or disabled.
     * @param enabled true to enable, false to disable
     */
    public void setSwipeEnabled(final boolean enabled) {
        mTouchHelperCallback.setSwipeEnabled(enabled);
    }

    /**
     * Sets long press drag gesture enabled or disabled.
     * @param enabled true to enable, false to disable
     */
    public void setLongPressDragEnabled(final boolean enabled) {
        mTouchHelperCallback.setLongPressDragEnabled(enabled);
    }

    /**
     * Sets manual drag gesture enabled or disabled.
     * @param enabled true to enable, false to disable
     */
    public void setManualDragEnabled(final boolean enabled) {
        mTouchHelperCallback.setManualDragEnabled(enabled);
    }

    /**
     * Returns true if swipe is enabled, false if swipe is disabled.
     * @return swipe state
     */
    public boolean isSwipeEnabled() {
        return mTouchHelperCallback.isItemViewSwipeEnabled();
    }

    /**
     * Returns true if long press drag is enabled, false if long press drag is disabled.
     * @return long press drag state
     */
    public boolean isLongPressDragEnabled() {
        return mTouchHelperCallback.isLongPressDragEnabled();
    }

    /**
     * Returns true if manual drag is enabled, false if manual drag is disabled.
     * @return manual drag state
     */
    public boolean isManualDragEnabled() {
        return mTouchHelperCallback.isManualDragEnabled();
    }

    /**
     * Class that builds {@link GestureManager} instance.
     */
    public static class Builder{

        private static final int INVALID_FLAG = -1;

        private RecyclerView recyclerView;

        private int swipeFlags = INVALID_FLAG;
        private int dragFlags = INVALID_FLAG;
        private boolean isSwipeEnabled;
        private boolean isDragEnabled;
        private boolean isManualDragEnabled;

        /**
         * Constructs {@link GestureManager} for the given RecyclerView.
         * @param recyclerView RecyclerView instance
         */
        public Builder(@NonNull final RecyclerView recyclerView) {
            this.recyclerView = recyclerView;
        }

        /**
         * Sets swipe gesture enabled or disabled.
         * Swipe is disabled by default.
         * @param enabled true to enable, false to disable
         * @return returns builder instance
         */
        public Builder setSwipeEnabled(final boolean enabled) {
            isSwipeEnabled = enabled;
            return this;
        }

        /**
         * Sets long press drag gesture enabled or disabled.
         * Long press drag is disabled by default.
         * @param enabled true to enable, false to disable
         * @return returns builder instance
         */
        public Builder setLongPressDragEnabled(final boolean enabled) {
            isDragEnabled = enabled;
            return this;
        }

        /**
         * Sets manual drag gesture enabled or disabled.
         * Manual drag is disabled by default.
         * @param enabled true to enable, false to disable
         * @return returns builder instance
         */
        public Builder setManualDragEnabled(final boolean enabled) {
            isManualDragEnabled = enabled;
            return this;
        }

        /**
         * Sets flags for swipe and drag gesture. Do not set this flags if you want predefined flags for RecyclerView layout manager.
         * See {@link ItemTouchHelper} flags.
         *
         * This method is deprecated, use {@link #setDragFlags(int)} or {@link #setSwipeFlags(int)}.
         * @param swipeFlags flags for swipe gesture
         * @param dragFlags flags for drag gesture
         * @return returns builder instance
         */
        @Deprecated
        public Builder setGestureFlags(final int swipeFlags, final int dragFlags) {
            this.swipeFlags = swipeFlags;
            this.dragFlags = dragFlags;
            return this;
        }

        /**
         * Sets flags for swipe gesture. Do not set this flags if you want predefined flags for RecyclerView layout manager.
         * See {@link ItemTouchHelper} flags.
         * @param flags flags for swipe gesture
         * @return returns builder instance
         */
        public Builder setSwipeFlags(final int flags) {
            swipeFlags = flags;
            return this;
        }

        /**
         * Sets flags for drag gesture. Do not set this flags if you want predefined flags for RecyclerView layout manager.
         * See {@link ItemTouchHelper} flags.
         * @param flags flags for drag gesture
         * @return returns builder instance
         */
        public Builder setDragFlags(final int flags) {
            dragFlags = flags;
            return this;
        }

        /**
         * Builds {@link GestureManager} instance.
         * @return returns GestureManager instance
         */
        public GestureManager build() {
            validateBuilder();
            return new GestureManager(this);
        }

        private void validateBuilder() {
            final boolean hasAdapter = recyclerView.getAdapter() instanceof GestureAdapter;
            if (!hasAdapter) {
                throw new IllegalArgumentException("RecyclerView does not have adapter that extends " + GestureAdapter.class.getName());
            }

            if (swipeFlags == INVALID_FLAG || dragFlags == INVALID_FLAG) {
                if (recyclerView.getLayoutManager() == null) {
                    throw new IllegalArgumentException("No layout manager for RecyclerView. Provide custom flags or attach layout manager to RecyclerView.");
                }
            }
        }
    }

}

File: library/src/main/java/com/thesurix/gesturerecycler/GestureManager.kt
package com.thesurix.gesturerecycler

import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper

private const val INVALID_FLAG = -1

/**
 * Class that is responsible for gesture management for [RecyclerView] widget.
 * @author thesurix
 */
class GestureManager {

    private val touchHelperCallback: GestureTouchHelperCallback

    private constructor(builder: Builder) {
        val adapter = builder.recyclerView.adapter as GestureAdapter<*, *>
        touchHelperCallback = GestureTouchHelperCallback(adapter).apply {
            swipeEnabled = builder.isSwipeEnabled
            longPressDragEnabled = builder.isDragEnabled
            manualDragEnabled = builder.isManualDragEnabled
        }

        val touchHelper = ItemTouchHelper(touchHelperCallback)
        touchHelper.attachToRecyclerView(builder.recyclerView)
        adapter.setGestureListener(GestureListener(touchHelper))

        if (builder.swipeFlags == INVALID_FLAG) {
            touchHelperCallback.setSwipeFlagsForLayout(builder.recyclerView.layoutManager!!)
        } else {
            touchHelperCallback.swipeFlags = builder.swipeFlags
        }

        if (builder.dragFlags == INVALID_FLAG) {
            touchHelperCallback.setDragFlagsForLayout(builder.recyclerView.layoutManager!!)
        } else {
            touchHelperCallback.dragFlags = builder.dragFlags
        }
    }

    /**
     * Returns true if swipe is enabled, false if swipe is disabled.
     * @return swipe state
     */
    /**
     * Sets swipe gesture enabled or disabled.
     * @param enabled true to enable, false to disable
     */
    var isSwipeEnabled: Boolean
        get() = touchHelperCallback.isItemViewSwipeEnabled
        set(enabled) {
            touchHelperCallback.swipeEnabled = enabled
        }

    /**
     * Returns true if long press drag is enabled, false if long press drag is disabled.
     * @return long press drag state
     */
    /**
     * Sets long press drag gesture enabled or disabled.
     * @param enabled true to enable, false to disable
     */
    var isLongPressDragEnabled: Boolean
        get() = touchHelperCallback.isLongPressDragEnabled
        set(enabled) {
            touchHelperCallback.longPressDragEnabled = enabled
        }

    /**
     * Returns true if manual drag is enabled, false if manual drag is disabled.
     * @return manual drag state
     */
    /**
     * Sets manual drag gesture enabled or disabled.
     * @param enabled true to enable, false to disable
     */
    var isManualDragEnabled: Boolean
        get() = touchHelperCallback.manualDragEnabled
        set(enabled) {
            touchHelperCallback.manualDragEnabled = enabled
        }

    /**
     * Class that builds [GestureManager] instance.
     * Constructs [GestureManager] for the given RecyclerView.
     * @param recyclerView RecyclerView instance
     */
    class Builder(val recyclerView: RecyclerView) {
        internal var swipeFlags = INVALID_FLAG
            private set
        internal var dragFlags = INVALID_FLAG
            private set
        internal var isSwipeEnabled = false
            private set
        internal var isDragEnabled = false
            private set
        internal var isManualDragEnabled = false
            private set

        /**
         * Sets swipe gesture enabled or disabled.
         * Swipe is disabled by default.
         * @param enabled true to enable, false to disable
         * @return returns builder instance
         */
        fun setSwipeEnabled(enabled: Boolean): Builder {
            isSwipeEnabled = enabled
            return this
        }

        /**
         * Sets long press drag gesture enabled or disabled.
         * Long press drag is disabled by default.
         * @param enabled true to enable, false to disable
         * @return returns builder instance
         */
        fun setLongPressDragEnabled(enabled: Boolean): Builder {
            isDragEnabled = enabled
            return this
        }

        /**
         * Sets manual drag gesture enabled or disabled.
         * Manual drag is disabled by default.
         * @param enabled true to enable, false to disable
         * @return returns builder instance
         */
        fun setManualDragEnabled(enabled: Boolean): Builder {
            isManualDragEnabled = enabled
            return this
        }

        /**
         * Sets flags for swipe and drag gesture. Do not set this flags if you want predefined flags for RecyclerView layout manager.
         * See [ItemTouchHelper] flags.
         *
         * This method is deprecated, use [.setDragFlags] or [.setSwipeFlags].
         * @param swipeFlags flags for swipe gesture
         * @param dragFlags flags for drag gesture
         * @return returns builder instance
         */
        @Deprecated("Use setSwipeFlags() and setDragFlags() methods.")
        fun setGestureFlags(swipeFlags: Int, dragFlags: Int): Builder {
            this.swipeFlags = swipeFlags
            this.dragFlags = dragFlags
            return this
        }

        /**
         * Sets flags for swipe gesture. Do not set this flags if you want predefined flags for RecyclerView layout manager.
         * See [ItemTouchHelper] flags.
         * @param flags flags for swipe gesture
         * @return returns builder instance
         */
        fun setSwipeFlags(flags: Int): Builder {
            swipeFlags = flags
            return this
        }

        /**
         * Sets flags for drag gesture. Do not set this flags if you want predefined flags for RecyclerView layout manager.
         * See [ItemTouchHelper] flags.
         * @param flags flags for drag gesture
         * @return returns builder instance
         */
        fun setDragFlags(flags: Int): Builder {
            dragFlags = flags
            return this
        }

        /**
         * Builds [GestureManager] instance.
         * @return returns GestureManager instance
         */
        fun build(): GestureManager {
            validateBuilder()
            return GestureManager(this)
        }

        private fun validateBuilder() {
            val hasAdapter = recyclerView.adapter is GestureAdapter<*, *>
            if (!hasAdapter) {
                throw IllegalArgumentException("RecyclerView does not have adapter that extends " + GestureAdapter::class.java.name)
            }

            if (swipeFlags == INVALID_FLAG || dragFlags == INVALID_FLAG) {
                if (recyclerView.layoutManager == null) {
                    throw IllegalArgumentException("No layout manager for RecyclerView. Provide custom flags or attach layout manager to RecyclerView.")
                }
            }
        }
    }

}
